File: backend/app/schemas/study_plan.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class StudyPlanCreate(BaseModel):
    goal: str = Field(..., min_length=2, max_length=500)
    level: str = Field(default="beginner", pattern=r"^(beginner|intermediate|advanced)$")
    duration_days: int = Field(..., ge=1, le=90)


class PlanTaskResponse(BaseModel):
    id: uuid.UUID
    day_index: int
    title: str
    description: str
    completed: bool
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class StudyPlanResponse(BaseModel):
    id: uuid.UUID
    goal: str
    level: str
    duration_days: int
    status: str
    created_at: datetime
    tasks: list[PlanTaskResponse] = []

    model_config = {"from_attributes": True}


class PlanProgressResponse(BaseModel):
    plan_id: uuid.UUID
    total_tasks: int
    completed_tasks: int
    progress_percent: float
    today_tasks: list[PlanTaskResponse] = []


class TaskCompleteRequest(BaseModel):
    completed: bool = True

File: mobile/lib/features/study_plan/models/study_plan.dart
class StudyPlan {
  StudyPlan({
    required this.id,
    required this.goal,
    required this.level,
    required this.durationDays,
    required this.status,
    required this.createdAt,
    this.tasks = const [],
  });

  factory StudyPlan.fromJson(Map<String, dynamic> json) {
    return StudyPlan(
      id: json['id'] as String,
      goal: json['goal'] as String,
      level: json['level'] as String,
      durationDays: json['duration_days'] as int,
      status: json['status'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      tasks: (json['tasks'] as List<dynamic>?)
              ?.map((t) => PlanTask.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  final String id;
  final String goal;
  final String level;
  final int durationDays;
  final String status;
  final DateTime createdAt;
  final List<PlanTask> tasks;
}

class PlanTask {
  PlanTask({
    required this.id,
    required this.dayIndex,
    required this.title,
    required this.description,
    required this.completed,
    this.completedAt,
  });

  factory PlanTask.fromJson(Map<String, dynamic> json) {
    return PlanTask(
      id: json['id'] as String,
      dayIndex: json['day_index'] as int,
      title: json['title'] as String,
      description: json['description'] as String? ?? '',
      completed: json['completed'] as bool? ?? false,
      completedAt: json['completed_at'] != null ? DateTime.parse(json['completed_at'] as String) : null,
    );
  }

  final String id;
  final int dayIndex;
  final String title;
  final String description;
  final bool completed;
  final DateTime? completedAt;
}

class PlanProgress {
  PlanProgress({
    required this.planId,
    required this.totalTasks,
    required this.completedTasks,
    required this.progressPercent,
    this.todayTasks = const [],
  });

  factory PlanProgress.fromJson(Map<String, dynamic> json) {
    return PlanProgress(
      planId: json['plan_id'] as String,
      totalTasks: json['total_tasks'] as int,
      completedTasks: json['completed_tasks'] as int,
      progressPercent: (json['progress_percent'] as num).toDouble(),
      todayTasks: (json['today_tasks'] as List<dynamic>?)
              ?.map((t) => PlanTask.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  final String planId;
  final int totalTasks;
  final int completedTasks;
  final double progressPercent;
  final List<PlanTask> todayTasks;
}
